javascript-manual/modules/ROOT/pages/query-async.adoc

```diff
@@ -0,0 +1,218 @@
+----
+
+<1> Run a query
+<2> Specify callback for successful runs, taking query result as input
+<3> Specify callback for failed runs, taking driver error as input
+<4> Specify callback to run regardless of query outcome
+
+[TIP]
+====
+The default return value of `tx.run()` is a `Promise`.
+If you prepend `await` to `tx.run()` as shown in xref:query-simple.adoc[Query the database], you effectively force your application to wait until the result is ready, and you obtain the corresponding of either `result` or `error` from the example above.
+====
+
+
+=== Combine multiple transactions
+
+To run multiple queries within the same transaction, use `Promise.all()`.
+It runs asynchronous operations concurrently, so you can submit multiple queries at the same time and wait for them all to finish.
+
+.Retrieve people's names and assign each of them to the company Neo4j
+[source, javascript]
+----
+const companyName = 'Neo4j'
+const session = driver.session({database: 'neo4j'})
+try {
+  const names = await session.executeRead(async tx => {
+    const result = await tx.run('MATCH (p:Person) RETURN p.name AS name')
+    return result.records.map(record => record.get('name'))
+  })
+
+  const relationshipsCreated = await session.executeWrite(tx =>
+    Promise.all(  // group together all Promises
+      names.map(name =>
+        tx.run(`
+          MATCH (emp:Person {name: $personName})
+          MERGE (com:Company {name: $companyName})
+          MERGE (emp)-[:WORKS_FOR]->(com)
+          `, { personName: name, companyName: companyName }
+        )
+        .then(result => result.summary.counters.updates().relationshipsCreated)
+      )
+    ).then(values => values.reduce((a, b) => a + b))  // aggregate results
+  )
+
+  console.log(`Created ${relationshipsCreated} employees relationships.`)
+} finally {
+  await session.close()
+}
+----
+
+
+== Streaming API
+
+The Streaming API allows you to run a query and receive results individually, as soon as the server has them ready.
+You can specify a callback to process each record.
+This API is particularly fit for cases in which it may take the server a different time to retrieve the different records, but you want to process each of them as soon as they are available.
+
+[source, javascript, role=nocollapse]
+----
+const session = driver.session({database: 'neo4j'})
+const records = session.executeWrite(tx => {
+  let peopleRecords = []
+  tx
+  .run('MERGE (p:Person {name: $name}) RETURN p.name AS name', {  // <1>
+    name: 'Alice'
+  })
+  .subscribe({  // <2>
+    onKeys: keys => {  // <3>
+      console.log('Result columns are:')
+      console.log(keys)
+    },
+    onNext: record => {  // <4>
+      console.log(`Processing ${record.get('name')}`)
+      peopleRecords.push(record)
+    },
+    onCompleted: () => {  // <5>
+      session.close() // returns a Promise
+    },
+    onError: error => {  // <6>
+      console.log(error)
+    }
+  })
+  return peopleRecords
+})
```

Comment:
This example is totally broken, the transaction can be committed before the return get streamed or the session can be closed before the transaction be committed. 

This example should be re-write as:

```suggestion
const peopleNames = await session.executeWrite(tx => new Promise((resolve, reject) => {
  let names = []
  tx
  .run('MERGE (p:Person {name: $name}) RETURN p.name AS name', {  // <1>
    name: 'Alice'
  })
  .subscribe({  // <2>
    onKeys: keys => {  // <3>
      console.log('Result columns are:')
      console.log(keys)
    },
    onNext: record => {  // <4>
      console.log(`Processing ${record.get('name')}`)
      names.push(record.get('name'))
    },
    onCompleted: () => {  // <5>
      resolve(names) // returns a Promise
    },
    onError: error => {  // <6>
      reject(error)
    }
  })
)

await session.close()
```

Reply:
I think all the Promise gymnastic one has to do to use the transaction function doesn't make the driver look too well, so I preferred to not use transaction functions in that example.